File: Labworks/Lab 1/GCD.c
#include <stdio.h>
#include <time.h>

int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int main() {
    int a, b;
    printf("Enter two numbers to find their GCD: ");
    scanf("%d %d", &a, &b);

    clock_t start_declare_time = clock();

    clock_t start_process_time = clock();

    int result = gcd(a, b);

    clock_t end_process_time = clock();

    printf("\nGCD of %d and %d is: %d\n", a, b, result);

    double declare_to_process_time = ((double)(start_process_time - start_declare_time)) / CLOCKS_PER_SEC * 1000;
    double process_duration = ((double)(end_process_time - start_process_time)) / CLOCKS_PER_SEC * 1000;
    double total_time = ((double)(end_process_time - start_declare_time)) / CLOCKS_PER_SEC * 1000;

    printf("\nTime taken from declaration to process start: %.6f ms\n", declare_to_process_time);
    printf("Time taken for the computation process: %.6f ms\n", process_duration);
    printf("Total time taken: %.6f ms\n", total_time);

    return 0;
}

File: Labworks/Lab 1/GCD.py
import time

def gcd(a, b):
    """Compute the GCD of two numbers using the Euclidean Algorithm."""
    while b:
        a, b = b, a % b
    return a

def main():
    print("Enter two numbers to compute their GCD:")
    num1 = int(input("First number: "))
    num2 = int(input("Second number: "))

    start_declare_time = time.time()
    start_process_time = time.time()

    result = gcd(num1, num2)

    end_process_time = time.time()

    print(f"\nThe GCD of {num1} and {num2} is: {result}")
    print(f"Time when variables were declared: {start_declare_time:.6f} seconds")
    print(f"Time when process started: {start_process_time:.6f} seconds")
    print(f"Time when process ended: {end_process_time:.6f} seconds")

    # calculate and display time taken
    declare_to_process_time = (start_process_time - start_declare_time) * 1000
    process_duration = (end_process_time - start_process_time) * 1000
    total_time = (end_process_time - start_declare_time) * 1000

    print(f"\nTime taken from declaration to process start: {declare_to_process_time:.6f} ms")
    print(f"Time taken for the computation process: {process_duration:.6f} ms")
    print(f"Total time taken: {total_time:.6f} ms")

if __name__ == "__main__":
    main()
